core/common/src/Utf8.kt

```diff
@@ -260,6 +260,9 @@ public fun Source.readString(byteCount: Long): String {
  */
 @OptIn(InternalIoApi::class)
 public fun Source.readCodePointValue(): Int {
+    if (this is Buffer) {
```

Comment:
@fzhinkin question, would this be a valid solution to avoid an `is` check?

```kotlin
public fun Buffer.readCodePointValue(): Int {
  return commonReadUtf8CodePoint()
}

public fun Source.readCodePointValue(): Int { 
  ... 
}
```

Reply:
No, it wouldn't, as extension functions are resolved statically based on the declared receiver's type.

So if there's a code like
```
val src: Source = Buffer().apply { ... }
println(src.readCodePointValue())
```
`readCodePointValue` will be resolved to `Source.readCodePointValue` and an extension defined on `Buffer` won't help.

Reply:
@fzhinkin thanks! If we have the additional extension function tho, it will avoid the check when the type is `Buffer`, which I think is the most common case.

Reply:
> it will avoid the check when the type is Buffer, which I think is the most common case.

Indeed, at least for Okio, the corresponding function is invoked on `Buffer` more frequently, compared to `BufferedSource`.

I don't really like the idea of adding an extra extension function, as performance benefits should be neglectingly small. On the other hand, there will be some benefits, so I consider adding the function later in the course of https://github.com/Kotlin/kotlinx-io/issues/342.

Reply:
Thanks! I think the best approach would be to do the same thing you did for unsafe operations, throw in a quick kotlinx-benchmark. Would be interesting to document all of these cases for JS.

For K/JS here we are lucky because Buffer is a class, and the `is` check will be compiled to a native `instanceof`. When the check is done on an interface, there is a lot more work involved.